tests/click_types.py

```diff
@@ -0,0 +1,163 @@
+                return (param_obj.path_type,)
+            else:
+                raise NotImplementedError(
+                    "todo: support the return type of a converter func"
+                )
+        else:
+            return (str,)
+
+    # globus-cli types
+    if isinstance(param_type, CommaDelimitedList):
+        return (list[str],)
+    elif isinstance(param_type, EndpointPlusPath):
+        if param_type.path_required:
+            return (tuple[uuid.UUID, str],)
+        else:
+            return (tuple[uuid.UUID, str | None],)
+    elif isinstance(param_type, IdentityType):
+        return (ParsedIdentity,)
+    elif isinstance(param_type, JSONStringOrFile):
+        return (JsonValue,)
+    elif isinstance(param_type, LocationType):
+        return (tuple[float, float],)
+    elif isinstance(param_type, StringOrNull):
+        return (str, ExplicitNullType)
+    elif isinstance(param_type, UrlOrNull):
+        return (str, ExplicitNullType)
+    elif isinstance(param_type, TaskPath):
+        return (TaskPath,)
+    elif isinstance(param_type, NotificationParamType):
+        return (dict[str, bool],)
+    elif isinstance(param_type, TimedeltaType):
+        if param_type._convert_to_seconds:
+            return (int,)
+        return (datetime.timedelta,)
+
+    raise NotImplementedError(f"unsupported parameter type: {param_type}")
+
+
+_NEVER_NULL_CALLBACKS = (none_to_empty_dict,)
+
+
+def _option_defaults_to_none(o: click.Option) -> bool:
+    if o.default is not None:
+        return False
+    if o.callback is None:
+        return True
+    if o.callback in _NEVER_NULL_CALLBACKS:
+        return False
+    return True
+
+
+def check_has_correct_annotations_for_click_args(f):
+    hints = t.get_type_hints(f.callback)
+    param_map = {}
+    for param in f.params:
+        # skip params which do not get passed to the callback
+        if param.expose_value is False:
+            continue
+
+        param_map[param.name] = param
+
+    for name in param_map:
+        if name not in hints:
+            raise ValueError(f"expected parameter '{name}' was not in type hints")
+
+    for name, param_obj in param_map.items():
+        possible_types = set()
+        # only implicitly add NoneType to the types if the default is None
+        # some possible cases to consider:
+        #   '--foo' is a string with an automatic default of None
+        #   '--foo/--no-foo' is a bool flag with an automatic default of False
+        #   '--foo/--no-foo' is a bool flag with an explicit default of None
+        #   '--foo' is a count option with a default of 0
+        #   '--foo' uses a param type which converts None to a default value
+        if isinstance(param_obj, click.Option):
+            if _option_defaults_to_none(param_obj):
+                possible_types.add(None.__class__)
+
+        for param_type in _param2types(param_obj):
+            if param_obj.multiple:
+                param_type = list[(param_type,)]
```

Comment:
Seeing `list[(param_type,)]` is starting to reinforce that passing a tuple into a type annotation like this is the correct syntax.

Reply:
I thought it was necessary, but after my little experiment on this, shown above, it might be valid to write `list[param_type]` as well.
If so, that might make me tend towards preferring unpacked tuples in general. `list[param_type]` reads better, I think.

Reply:
I was working through several changes here and I found that it was hard or at least fiddly to make things totally consistent.

`flake8` flags the following syntax as invalid (`E999`):
```
Literal[*tuple(choices)]
```

Now... it is not invalid, but I can understand why it looks that way. `__getitem__` takes a single arg, and I'm guessing the flake8 parser doesn't have updates to handle `__class_getitem__`.

So we could ignore this, but breaking a parser which is being used on our code seems unwise. So I'm ending up mixing styles here based on the usage.

`tuple[param_type, ...]` gets used in one place, `Literal[tuple(choices)]` in another.

I think this is okay, if a _little_ unfortunate, and it just needs to be part of our understanding of how runtime evaluation of types works.